### DISCUSSION TITLE
MacOS and Windows images not running tests

### DISCUSSION THREAD

**Discussion by druzm** [actions/virtual-environments#7744] [General questions] [answered]
Hi,

I'm not sure whether I should report this issue, but I think starting a discussion about it and seeing if other people are experiencing the same issue might help.

I work on a Python package that needs to be compatible with Linux (Ubuntu), MacOS, and Windows. Therefore, I have set up the CI accordingly. For about a year, everything worked well. I would push to the repository, and the CI would run tests for different Python versions and operating systems. However, approximately a month ago, the MacOS and Windows images never seem to finish running the tests. It doesn't matter whether I set the timeout to one or six hours; sometimes not even a single test is executed. However, everything continues to work with Ubuntu.

Here there's a link to one of the most recent runs: https://github.com/gaia-dpci/GaiaXPy/actions/runs/5289393980

I contacted GitHub support and was redirected here.

Thanks!

**Comment by vpolikarpov-akvelon** [comment]
Hi @druzm. I forked your repo and experimented a bit. I selected two subsequent commits from `polish-finder-release` branch and ran workflow for each of them several times. The results are:
- workflow always works with no problems for https://github.com/gaia-dpci/GaiaXPy/tree/8d93408ade
- windows and macOS jobs hang for https://github.com/gaia-dpci/GaiaXPy/tree/037cba7e84

Image versions were the same for both commits (`windows-2022 20230612.1`, `macos-12 20230612.1`). So it looks like something is wrong with your code but not runners.

**Reply by druzm** [reply]
Hi @vpolikarpov-akvelon. This issue began exactly or almost exactly on a day when Actions was experiencing degraded performance: https://twitter.com/githubstatus/status/1658582982775525376. On that day, it was impossible to run a single test remotely, but we continued to make changes in the code. Eventually, everything started working again with Ubuntu, but not with the rest of the images.

I found your findings interesting, so I decided to run the tests locally on Windows (I have dual boot). The tests ran without any issues, but since I ran all the tests, including those that generate plots displayed on a window (which I usually skip locally), I thought of trying to skip the plot-generating tests once more (although I had tried this before without success). Surprisingly, this time the tests passed for macOS and Windows, except for macOS with Python 3.7, which failed with `ModuleNotFoundError: No module named '_bz2'`.

Now I believe the issue probably has/had to do with these new tests that display plots, which were added from a collaborator's fork. My assumption is that the runner generates the plot, opens a window no one can see, and because the window is never closed, it gets stuck there and eventually times out. As I mentioned, sometimes no tests would be executed at all, and other times the CI would get stuck. So, this is not the first time I've tried skipping these tests. However, for some reason, the tests are passing now.

Do you know why running these types of tests that open plots is not a problem with Ubuntu, but it is with macOS and Windows? I wonder whether it may have to do with the GUI backend.

Thanks!

**Reply by vpolikarpov-akvelon** [reply]
I created machine from runner image and tried to run your workflow manually. Well, you are right: tests open windows that never get closed so execution hangs. Here is how it looks on Windows:
![image](https://github.com/actions/runner-images/assets/126792224/e3cd781d-001f-4747-95e0-6f3d31fddb51)

I bet the same is on macOS. The reason for your workflow to work in Ubuntu is the absence of graphical UI on Ubuntu runners.

And speaking of `ModuleNotFoundError`, it's a known issue with python 3.7.17 that you can track here: https://github.com/actions/setup-python/issues/682

**Reply by druzm** [reply]
At the moment, I've opted for skipping those tests by simply running a custom script that ignores the corresponding folder, but I'm not sure this is ideal. Are there any other alternatives you know of, like suppressing the graphical UI for the affected OSs, for example?

**Reply by vpolikarpov-akvelon** [reply]
As far as I know there is no easy solution for that. In general you have two options:
1. Write some script that would work in background and close all those windows
2. Restrict opening windows using group policy (on Windows) or remove execution permission from the binary that opens windows (on macOS)

Both options are quite touch to implement and are unreliable from my point of view. I think it would be easier to change tests somehow.

**Reply by druzm** [reply, marked as answer]
I can try passing a flag to my tests script, so I can save the plots without displaying them. It's not like anyone will see them either, but I can choose to display them when running the tests locally and not to when running them remotely.
Thanks!
